File: app/src/main/java/com/xiaomai/geek/module/base/BaseActivity.java

package com.xiaomai.geek.module.base;

import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import com.trello.rxlifecycle.LifecycleTransformer;
import com.trello.rxlifecycle.components.support.RxAppCompatActivity;
import com.xiaomai.geek.R;
import com.xiaomai.geek.presenter.IBasePresenter;
import com.xiaomai.geek.view.IBaseView;
import com.xiaomai.geek.wedget.EmptyView;

import javax.annotation.Nullable;
import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by XiaoMai on 2017/3/24 15:01.
 * 
 * @param <T>
 */

public abstract class BaseActivity<T extends IBasePresenter> extends RxAppCompatActivity
        implements IBaseView {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(attachLayoutRes());
        ButterKnife.bind(this);
        initInjector();
        initViews();
        initSwipeRefresh();
        updateViews(false);
    }

    /**
     * 把 EmptyLayout 放在基类统一处理，
     * 
     * @Nullable 表明 View 可以为 null，详细可看 ButterKnife
     */
    @Nullable
    @BindView(R.id.layout_empty)
    protected EmptyView mEmptyView;

    /**
     * 把 Presenter 提取到基类需要配合基类的 initInjector() 进行注入，</br>
     * 如果继承这个基类则必定要提供一个 Presenter 注入方法， 该 APP 所有 Presenter 都是在 Module
     * 提供注入实现，也可以选择提供另外不带 Presenter 的基类
     */
    @Inject
    protected T mPresenter;

    @Nullable
    // @BindView(R.id.swip_refresh)
    SwipeRefreshLayout mSwipeRefresh;

    /**
     * 绑定布局文件
     * 
     * @return 布局文件id
     */
    @LayoutRes
    protected abstract int attachLayoutRes();

    /**
     * Dagger 注入
     */
    protected abstract void initInjector();

    protected abstract void initViews();

    protected abstract void updateViews(boolean isRefresh);

    private void initSwipeRefresh() {
        if (mSwipeRefresh != null) {
            // SwipeRefreshH
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void showLoading() {
        if (mEmptyView != null) {
            mEmptyView.setStatus(EmptyView.STATUS_LOADING);
        }
    }

    @Override
    public void disMissLoading() {
        if (mEmptyView != null)
            mEmptyView.hide();
    }

    @Override
    public void showError(EmptyView.OnRetryListener onRetryListener) {
        if (mEmptyView != null) {
            mEmptyView.setStatus(EmptyView.STATUS_ERROR);
            mEmptyView.setOnRetryListener(onRetryListener);
        }
    }

    @Override
    public void finishRefresh() {
        if (mSwipeRefresh != null) {
            mSwipeRefresh.setRefreshing(false);
        }
    }

    protected void initToolBar(Toolbar toolbar, boolean homeAsUpEnable, String title) {
        toolbar.setTitle(title);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null)
            actionBar.setDisplayHomeAsUpEnabled(homeAsUpEnable);
    }

    protected void initToolBar(Toolbar toolbar, boolean homeAsUpEnable, int resTitle) {
        initToolBar(toolbar, homeAsUpEnable, getString(resTitle));
    }

    @Override
    public <T> LifecycleTransformer<T> bindToLife() {
        return this.bindToLifecycle();
    }
}

File: app/src/main/java/com/xiaomai/geek/base/BaseActivity.kt
package com.xiaomai.geek.base

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.v7.app.AppCompatActivity
import android.view.LayoutInflater
import android.view.View
import com.xiaomai.geek.R
import kotlinx.android.synthetic.main.geek_base_activity.*
import kotlinx.android.synthetic.main.geek_empty_view.*
import kotlinx.android.synthetic.main.geek_error_view.*

/**
 * Created by wangce on 2018/1/30.
 */
abstract class BaseActivity : AppCompatActivity() {
    val TAG = javaClass.simpleName

    lateinit var dataBinding: ViewDataBinding

    open fun showLoading() {
        swipe_refresh_layout.isRefreshing = true
    }

    open fun dismissLoading() {
        swipe_refresh_layout.isRefreshing = false
    }

    open fun showContent() {
        resetPage(swipe_refresh_layout)
    }

    open fun showEmpty() {
        resetPage(empty_root_layout)
    }

    open fun showError() {
        resetPage(error_root_layout)
    }

    fun resetPage(view: View) {
        swipe_refresh_layout.isRefreshing = false
        swipe_refresh_layout.visibility = View.GONE
        empty_root_layout.visibility = View.GONE
        error_root_layout.visibility = View.GONE

        view.visibility = View.VISIBLE
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (useBaseLayout()) {
            DataBindingUtil.setContentView<ViewDataBinding>(this, R.layout.geek_base_activity)
            if (getLayoutId() > 0) {
                dataBinding = DataBindingUtil.inflate<ViewDataBinding>(LayoutInflater.from(this@BaseActivity), getLayoutId(), content_view, true)
//                LayoutInflater.from(this@BaseActivity).inflate(getLayoutId(), content_view, true)
            }
        } else {
            dataBinding = DataBindingUtil.setContentView<ViewDataBinding>(this, getLayoutId())
//            setContentView(getLayoutId())
        }
    }

    @LayoutRes
    abstract fun getLayoutId(): Int

    open fun useBaseLayout() = true
}